fix: Improve web-ext start-up time by lazy loading some modules

Rob--W opened the pull request:
Simple commands like `web-ext --help` now require less than a second instead of 4 seconds to complete.

And `web-ext run` also runs slightly over 3 seconds faster (https://github.com/mozilla/web-ext/pull/1302#issuecomment-380786023)

Fixes #1301

rpl commented on src/cmd/index.js:
```diff
@@ -1,8 +1,12 @@
+
+for (const cmd of ['build', 'lint', 'run', 'sign', 'docs']) {
+  defaultCommands[cmd] = async function(params: Object, options: Object) {
+    const {default: runCommand} = await import(`./${cmd}`);
```
To be fair I'm not really thrilled about the webpack generated bundle chunks due to the dynamic import, especially numbered.

if we have to generate more then one transpiled file, it could be reasonable to just transpile everything file by file by using something like `babel-cli -d lib src/`, that at least would be less confusing (but it also would require a number of other changes, and I would prefer to avoid that for now).

Another option, if we agree on splitting the webpack bundle into chunks, would be to use set the webpack `output.chunkFilename` config option (https://webpack.js.org/configuration/output/#output-chunkfilename) to something like 'web-ext.[name].js' and then annotate the import using webpackChunkName webpack inline comment (https://webpack.js.org/api/module-methods/#import-), which would be possible only if paired with the more explicit approach described above, something like:

```js
export default {
  async build(
    params: BuildCmdParams, options: BuildCmdOptions
  ): Promise<ExtensionBuildResult> {
    const {default: cmd} = await import(
      /* webpackChunkName: "cmd-build" */
      './build');
    return cmd(params, options);
  },
```

One last option could be the "webpackMode" set to "eager", which would prevent webpack from generating any chunk:

```
for (const cmd of ['build', 'lint', 'run', 'sign', 'docs']) {
  defaultCommands[cmd] = async function(
    params: Object, options: Object
  ): Promise<any> {
    // $FLOW_FIXME: flow complains about any non literal module import or require.
    const {default: runCommand} = await import(
      /* webpackMode: "eager" */
      `./${cmd}`);
    return runCommand(params, options);
  };
}
```

which would be fine if we are ok with leave the type annotations of the functions exported by this module that open (I mean that `Object` is kind of "catch all" for the flow type checking, similarly to `Function` and `any`).

To be fair it doesn't matter when this module is used as an exported API to be used in other projects (because all the flow type are stripped when the sources are transpiled), and the type checking of the "program.js" module file is already pretty "relaxed" (but lets also take a look at the flow coverage more explicitly to evaluate the impact of it, one way to do it is using [flow-coverage-report](https://github.com/rpl/flow-coverage-report/) ;-))

rpl commented:
@Rob--W there is also another issue that I've just noticed: the way we currently instrument the sources and then collect the code coverage data from the test units doesn't like the dynamic import.

The simpler way to fix it (without changing anything on the "instrumenting and collecting the code coverage data" side, yet) seems to be to use the nodejs `require` instead of the dynamic `import` (which also means that the webpack bundle will not be chunked and we doesn't need to configure webpack to name the chunks or to force webpack to do not create any chunks for the dynamic imports), at least temporarily as a first step.

Another way to fix that (and use the dynamic import) would be to "update" our code coverage instrumentation to use `babel-plugin-istanbul` instead of `babel-istanbul-loader`, then the code coverage data collection and reporting to nyc, and also rework the way we run the tests (basically moving away from the grunt mochaTest task, and just call nyc and mocha directly from the npm scripts).

Once we fix the code coverage reporting, I'm pretty sure that coveralls will complain at least about the new code in the `src/cmd/index.js` not being covered and the total coverage going down as a side effect of that.

Using `require` instead on `import` currently seems to be the option that would allow us to introduce this big performance improvement with the minimum amount of upfront changes, and it would still allow us to introduce the other ones (e.g. use import instead of require, split the transpiled code in smaller files, prepare to remove some old package like babel-istanbul-loader, and maybe even grunt at some point).

Let me know how that sounds to you.

rpl commented on tests/unit/test.web-ext.js:
```diff
@@ -14,10 +13,25 @@ describe('webExt', () => {
+      try {
+        await runCommand(params, options);
+      } finally {
+        stub.restore();
```
(same for the in tests/unit/test-cmd/test.run.js, if we opt to keep the lazy loading of the extension runners as it is in this pull request).

Rob--W commented on src/cmd/index.js:
```diff
@@ -1,8 +1,47 @@
+async function build(
+  params: BuildCmdParams, options: BuildCmdOptions
+): Promise<ExtensionBuildResult> {
+  const {default: runCommand} = require('./build');
```
I've created an issue at #1306 and will reference it in the comments.

Rob--W commented:
I have made the following changes:

- (Trivial) Add comments to first patch: https://pastebin.mozilla.org/9083059
- Rewritten the test.web-ext.js test to use `afterEach` instead of try-catch: https://pastebin.mozilla.org/9083062
- Moved the `require` from `cmd/run.js` to `extension-runners/index.js` (and refactored a test to avoid having to pass in a `FirefoxDesktopExtensionRunner` or `FirefoxAndroidExtensionRunner`): https://pastebin.mozilla.org/9083068
